### PR TITLE
[FLOC-1542] Split container discovery

### PR DESCRIPTION
**Pull request by itamarst**
Will be more readable once #1233 is merged (which can be done as soon as buildbot is green).

### CODE REVIEW

File: flocker/node/_deploy.py [hunk 1]
```diff
@@ -410,9 +415,30 @@ def run(self, deployer):
...
+    """
+    Discover and calculate changes for peer-to-peer manifestations
+    (e.g. ZFS) on a node.
+
+    :ivar unicode hostname: The hostname of the node that this is running
+            on.
+    :ivar VolumeService volume_service: The volume manager for this node.
+    """
+    def discover_local_state(self, local_state):
+        # XXX Move manifestation discovery code out of ApplicationNodeDeployer,
```
Comment (exarkun) [inline flocker/node/_deploy.py]: including path information, I suppose

File: flocker/node/_deploy.py [hunk 2]
```diff
@@ -816,3 +850,30 @@ def find_dataset_changes(hostname, current_state, desired_state):
...
+
+    Temporary expedient, to be removed in FLOC-1553 or perhaps another
+    sub-task of FLOC-1443.
+    """
+    def __init__(self, hostname, volume_service, docker_client=None,
+                 network=None):
+        self.manifestations_deployer = ...
+        self.applications_deployer = ...
+
+    def discover_local_state(self, cluster_state):
```
Comment (exarkun) [inline flocker/node/_deploy.py]: in the interface, the new argument is `local_state`

File: flocker/node/_deploy.py [hunk 3]
```diff
@@ -80,11 +80,17 @@ class IDeployer(Interface):
     changes to bring local state and desired cluster configuration into
     alignment.
     """
-    def discover_local_state():
+    def discover_local_state(local_state):
```
Comment (exarkun) [inline flocker/node/_deploy.py]: This impacts the caller.  Can you add something about what changes this will require in `_loop.py`?

**Comment by exarkun** [comment]
Thanks.  Seems sensible.  Some inconsistencies, probably fine, but please address those and submit for one more quick look.

File: flocker/node/_loop.py [hunk 4]
```diff
@@ -275,14 +275,15 @@ def output_STORE_INFO(self, context):
...
+        # (i.e. we know nothing).
+        d = DeferredContext(self.deployer.discover_local_state(
+            known_local_state))
 
         def got_local_state(local_state):
             # Current cluster state is likely out of date as regards the local
             # state, so update it accordingly.
-            # XXX FLOC-1542 will need to deal with partial updates, and
-            # make sure there is no duplication with code in
-            # flocker.control._clusterstate.
```
Comment (exarkun) [inline flocker/node/_loop.py]: Something more necessary here?  Call `cluster_state.update_node` instead of `cluster_state.set`?

Pre-existing issue - the attribute is named `cluster_state` but the documentation for the class says `state` and that it is a `Deployment` but I think it's a `DeploymentState` now.

**Comment by exarkun** [comment]
Great, thanks.  Left one more question inline but I'm pretty sure I know the answer.  Please proceed.

File: flocker/node/test/test_deploy.py [hunk 5]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+        """
+        ``ApplicationNodeDeployer.discover_local_state`` returns ``NodeState``
+        with a a list of running ``Application``\ s; one for each active
+        container.
+        """
+        expected_application_name = u'site-example.com'
+        unit = Unit(name=expected_application_name,
+                    container_name=expected_application_name,
+                    container_image=u"flocker/wordpress:latest",
+                    activation_state=u'active')
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: This unit is repeated a lot of times in the following tests (often with minor variations).  Most or all of the duplication could be removed if `Unit` were switched to `pyrsistent` and a template `Unit` created that each test could either use directly or use to create its slightly different desired `Unit`.

File: flocker/node/test/test_deploy.py [hunk 6]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+        fake_docker = FakeDockerClient(units={expected_application_name: unit})
+        application = Application(
+            name=unit.name,
+            image=DockerImage.from_string(unit.container_image)
+        )
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: :question:

I started to write a comment about how an `application_from_unit` helper function would aid in removing some of the repetition of `Application` construction in these tests.  That led me to realize the point of most of these tests is exactly to test various edge cases in that function (if it existed).  Perhaps if it did, instead of the logic being embedded in the middle of `discover_node_state`, these tests would be easier to write.

File: flocker/node/test/test_deploy.py [hunk 7]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+        """
+        environment_variables = (
+            ('CUSTOM_ENV_A', 'a value'),
+            ('CUSTOM_ENV_B', 'something else'),
+        )
+        link_environment_variables = (
+            ('APACHE_PORT_80_TCP', 'tcp://example.com:8080'),
+            ('APACHE_PORT_80_TCP_PROTO', 'tcp'),
+            ('APACHE_PORT_80_TCP_ADDR', 'example.com'),
+            ('APACHE_PORT_80_TCP_PORT', '8080'),
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: `b".."`

File: flocker/node/test/test_deploy.py [hunk 8]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+
+    def test_discover_application_with_environment(self):
+        """
+        An ``Application`` with ``Environment`` objects is discovered from a
+        ``Unit`` with ``Environment`` objects.
+        """
+        environment_variables = (
+            ('CUSTOM_ENV_A', 'a value'),
+            ('CUSTOM_ENV_B', 'something else'),
+        )
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: `b".."`

File: flocker/node/test/test_deploy.py [hunk 9]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+        An ``Application`` with ``Link`` objects is discovered from a ``Unit``
+        with environment variables that correspond to an exposed link.
+        """
+        fake_docker = FakeDockerClient()
+        applications = [
+            Application(
+                name=u'site-example.com',
+                image=DockerImage.from_string(u'clusterhq/wordpress:latest'),
+                links=frozenset([
+                    Link(local_port=80, remote_port=8080, alias='APACHE')
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: `u".."`

File: flocker/node/test/test_deploy.py [hunk 10]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+            )
+        ]
+        api = ApplicationNodeDeployer(
+            u'example.com',
+            docker_client=fake_docker,
+            network=self.network
+        )
+        for app in applications:
+            StartApplication(
+                hostname='node1.example.com', application=app
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: `u".."` ... and I'm gonna stop commenting on this type of thing.

File: flocker/node/test/test_deploy.py [hunk 11]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+        ]
+        api = ApplicationNodeDeployer(
+            u'example.com',
+            docker_client=fake_docker,
+            network=self.network
+        )
+        d = api.discover_local_state(NodeState(hostname=u'example.com'))
+
+        self.assertItemsEqual(pset(applications),
+                              sorted(self.successResultOf(d).applications))
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: are `pset` and `sorted` actually helpful here?  `assertItemsEqual`'s job is to handle those details, isn't it?  Also comparing a set to a list is wonky.

File: flocker/node/test/test_deploy.py [hunk 12]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+            )
+        ]
+        api = ApplicationNodeDeployer(
+            u'example.com',
+            docker_client=fake_docker,
+            network=self.network
+        )
+        d = api.discover_local_state(NodeState(hostname=u'example.com'))
+
+        self.assertEqual(sorted(applications),
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: `assertItemsEqual`?

File: flocker/node/_deploy.py [hunk 13]
```diff
@@ -80,11 +80,20 @@ class IDeployer(Interface):
     changes to bring local state and desired cluster configuration into
     alignment.
     """
-    def discover_local_state():
+    hostname = Attribute(
+        "The hostname of the node this deployer is managing.")
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Please add type information to this documentation.

File: flocker/node/_deploy.py [hunk 14]
```diff
@@ -410,36 +418,20 @@ def run(self, deployer):
...
         self.volume_service = volume_service
 
-    def discover_local_state(self):
-        """
-        List all the ``Application``\ s running on this node.
-
-        :returns: A ``Deferred`` which fires with a ``NodeState``
-            instance.
-        """
+    def discover_local_state(self, local_state):
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Docstring

File: flocker/node/_deploy.py [hunk 15]
```diff
@@ -455,14 +447,92 @@ def map_volumes_to_size(volumes):
...
 
-        def applications_from_units(result):
-            units, available_manifestations = result
-            applications = []
+        def got_volumes(available_manifestations):
             manifestations = []  # Manifestation objects we're constructing
             manifestation_paths = {dataset_id: path for (path, (dataset_id, _))
                                    in available_manifestations.items()}
+
+            manifestations += list(
```
Comment (exarkun) [inline flocker/node/_deploy.py]: It was an empty list before this.  Just `manifestations = ...`, perhaps, and lose `manifestations = []` above?

File: flocker/node/_deploy.py [hunk 16]
```diff
@@ -455,14 +447,92 @@ def map_volumes_to_size(volumes):
...
+                manifestations={manifestation.dataset_id: manifestation
+                                for manifestation in manifestations},
+                paths=manifestation_paths,
+            )
+        volumes.addCallback(got_volumes)
+        return volumes
+
+    def calculate_necessary_state_changes(self, *args, **kwargs):
+        # Does nothing in this branch. Follow up will move
+        # calculate_necessary_state_changes code here.
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Add an issue link.

File: flocker/node/_deploy.py [hunk 17]
```diff
@@ -455,14 +447,92 @@ def map_volumes_to_size(volumes):
...
+        """
+        List all the ``Application``\ s running on this node.
+
+        The given local state is used to figure out if applications have
+        attached volumes that are specific manifestations.
+
+        :return: A ``Deferred`` which fires with a ``NodeState`` instance
+            with information only about
+            ``Application``. ``NodeState.manifestations`` and
+            ``NodeState.paths`` will not be filled in.
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Looks oddly mis-wrapped.

File: flocker/node/_deploy.py [hunk 18]
```diff
@@ -658,6 +714,10 @@ def calculate_necessary_state_changes(self, local_state,
         for application_name in applications_to_inspect:
             inspect_desired = desired_applications_dict[application_name]
             inspect_current = current_applications_dict[application_name]
+            # Current state never has metadata, but that's OK:
+            if inspect_desired.volume:
```
Comment (exarkun) [inline flocker/node/_deploy.py]: `... is not None:`

File: flocker/node/_deploy.py [hunk 19]
```diff
@@ -733,7 +793,13 @@ def change_node_state(deployer, desired_configuration,  current_cluster_state):
 
     :return: ``Deferred`` that fires when the necessary changes are done.
     """
-    d = deployer.discover_local_state()
+    nodes = [node for node in current_cluster_state.nodes
+             if node.hostname == deployer.hostname]
+    if nodes:
+        node = [nodes[0]]
```
Comment (exarkun) [inline flocker/node/_deploy.py]: This seems like it must be wrong (missing test case?).  `node = nodes[0]`?

File: flocker/node/test/test_deploy.py [hunk 20]
```diff
@@ -666,8 +669,7 @@ def test_does_not_exist(self):
 # XXX For now we require volume names match application names,
 # see https://github.com/ClusterHQ/flocker/issues/49
 DATASET_ID = unicode(uuid4())
-DATASET = Dataset(dataset_id=DATASET_ID,
-                  metadata=pmap({u"name": APPLICATION_WITH_VOLUME_NAME}))
+DATASET = Dataset(dataset_id=DATASET_ID)
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: That metadata wasn't there to cause any particular code paths to be exercised?  And it seems like the comment is out of date now (and perhaps has been for a while - since we switched to UUIDs for dataset ids, I suppose?).  The dataset no longer has a matching name nor a matching "name" metadata item and it seems the `ApplicationNodeDeployer` logic also doesn't care about either of these things (not sure when that changed).  If this is right, delete the comment.

File: flocker/node/test/test_deploy.py [hunk 21]
```diff
@@ -723,7 +729,7 @@ def test_discover_none(self):
             docker_client=fake_docker,
             network=self.network
         )
-        d = api.discover_local_state()
+        d = api.discover_local_state(NodeState(hostname=u"example.com"))
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: Lots of incredibly boring repetitive `NodeState` creation.  Refactor this.

File: flocker/node/test/test_deploy.py [hunk 22]
```diff
@@ -1050,7 +1055,6 @@ def test_discover_locally_owned_volume_with_size(self):
                     manifestation=Manifestation(
                         dataset=Dataset(
                             dataset_id=DATASET_ID,
-                            metadata=pmap({"name": unit1.name}),
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: This test method's docstring still makes claims about things with the "same name".

File: flocker/node/test/test_deploy.py [hunk 23]
```diff
@@ -1308,6 +1311,572 @@ def test_discover_manifestation_paths(self):
...
+                         sorted(self.successResultOf(d).applications))
+
+    def test_unknown_manifestations(self):
+        """
+        If the given ``NodeState`` indicates ignorance of manifestations, the
+        ``ApplicationNodeDeployer`` doesn't bother doing any discovery and
+        just indicates ignorance of applications.
+
+        XXX file issue for "it's not possible to run the container agent
+        without any dataset agents".
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: File that issue, I suppose.

File: flocker/node/test/test_deploy.py [hunk 24]
```diff
@@ -2112,6 +2700,71 @@ def test_volume_resized_before_move(self):
...
+            volume_service, docker_client=docker,
+            network=make_memory_network()
+        )
+
+        changes = api.calculate_necessary_state_changes(
+            self.successResultOf(api.discover_local_state(current_nodes[0])),
+            desired_configuration=desired,
+            current_cluster_state=current,
+        )
+        self.assertEqual(changes, Sequentially(changes=[]))
```
Comment (exarkun) [inline flocker/node/test/test_deploy.py]: :question:

I'm tempted to suggest that a better testing strategy here would be to define the cluster configuration, discover state, calculate changes, **run** the changes, discover state, calculate changes again - and then assert that the changes are null.  This has the benefit of running less fake code.  I'm somewhat afraid that it will turn out to have the major downside of being broken in some random irrelevant way, though.  Try it if you want, I guess.

**Comment by exarkun** [comment]
Thanks.  Please address inline comments and then merge.

**Comment by exarkun** [comment]
(Also green buildbot etc)

File: flocker/node/test/test_deploy.py (hunk 6 above)
Comment (itamarst) [inline flocker/node/test/test_deploy.py]: I ended up making corresponding shared `Application` instance, so that reduced duplication.

File: flocker/node/_deploy.py (hunk 19 above)
Comment (itamarst) [inline flocker/node/_deploy.py]: I suspect an acceptance test will catch this. Stupid legacy flocker-changestate stuff.

File: flocker/node/test/test_deploy.py (hunk 20 above)
Comment (itamarst) [inline flocker/node/test/test_deploy.py]: At some point we switched matching to be based on paths, not name, yeah.

File: flocker/node/test/test_deploy.py (hunk 22 above)
Comment (itamarst) [inline flocker/node/test/test_deploy.py]: This is the legacy testcase that is just there temporarily, so not going to be maintaining it.

File: flocker/node/test/test_deploy.py (hunk 24 above)
Comment (itamarst) [inline flocker/node/test/test_deploy.py]: Hm. I'm guessing splitting up calculation logic is going to have sufficient impact on these tests that it's not worth messing with them too much.
